sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala

```diff
@@ -204,5 +205,45 @@ object SchemaPruning extends Rule[LogicalPlan] {
     }
   }
 
+  private def applyMetadataSchemaPruning(plan: LogicalPlan): LogicalPlan =
+    plan transformDown {
+      case op @ PhysicalOperation(projects, filters, l @ LogicalRelation(_, _, _, _))
+        if containsMetadataAttributes(l) => pruneMetadataSchema(l, projects, filters).getOrElse(op)
+    }
+
+  /**
+   * This method returns optional logical plan with pruned metadata schema.
+   * `None` is returned if no nested field is required or all nested fields are required.
+   */
+  private def pruneMetadataSchema(
+      relation: LogicalRelation,
+      projects: Seq[NamedExpression],
+      filters: Seq[Expression]): Option[LogicalPlan] = {
+    val output = relation.output
+    val (normalizedProjects, normalizedFilters) =
+      normalizeAttributeRefNames(output, projects, filters)
+    val requestedRootFields = identifyRootFields(normalizedProjects, normalizedFilters)
+
+    val metadataSchema = output.collect { case MetadataAttribute(attr) => attr }.toStructType
+    val prunedMetadataSchema = pruneDataSchema(metadataSchema, requestedRootFields)
+
+    // If the metadata schema is different from the pruned metadata schema, continue.
+    // Otherwise, return None.
+    if (countLeaves(metadataSchema) > countLeaves(prunedMetadataSchema)) {
+      val projectionOverSchema = ProjectionOverSchema(prunedMetadataSchema)
+      Some(buildNewProjection(projects, normalizedProjects, normalizedFilters,
```

Comment:
Shall we update the scan node as well? It's a bit weird that the query plan has attributes with the same expr id but not the same data type.